### ISSUE TITLE
Broken codegen for apple menu events on macOS since wxFB 3.8

### ISSUE THREAD

**Issue by Zenju** [wxFormBuilder/wxFormBuilder#430]
Hi,

since wxFB 3.8 the menu events in the apple menu (= the menu named after the app that is running) are not working: Neither the keyboard shortcuts work, nor manually clicking with the mouse on the menu items.
![1](https://user-images.githubusercontent.com/7656530/42724103-8928c4f6-876c-11e8-9eae-afe79c611709.PNG)

This is due to the following code generation changes:

Before wxFB 3.8:
```
this->Connect( m_menuItemOptions->GetId(), wxEVT_COMMAND_MENU_SELECTED, wxCommandEventHandler( MainDialogGenerated::OnMenuOptions ) );
this->Connect( m_menuItemAbout->GetId(), wxEVT_COMMAND_MENU_SELECTED, wxCommandEventHandler( MainDialogGenerated::OnMenuAbout ) );
```
Since wxFB 3.8:
```
m_menuTools->Bind(wxEVT_COMMAND_MENU_SELECTED, wxCommandEventHandler( MainDialogGenerated::OnMenuOptions ), this, m_menuItemOptions->GetId());
m_menuHelp->Bind(wxEVT_COMMAND_MENU_SELECTED, wxCommandEventHandler( MainDialogGenerated::OnMenuAbout ), this, m_menuItemAbout->GetId());
```

The reason for the breakage seems to be that wxWidgets has special handling for the "Options" and "About" menu items on macOS. Even though they are designed to be located in my case under "menu->tools" and "menu->help" in wxFB, wxWidgets moves them out to the apple menu to satisfy the macOS human interface guidelines.

Obviously this means that the event is then originating from the apple menu, and NOT from the menus where they reside during design time in wxFB, m_menuTools and m_menuHelp. I can confirm this in my tests, i.e. the following code correctly triggers the events:

```
Bind(wxEVT_COMMAND_MENU_SELECTED, wxCommandEventHandler( MainDialogGenerated::OnMenuOptions ), this, m_menuItemOptions->GetId());
Bind(wxEVT_COMMAND_MENU_SELECTED, wxCommandEventHandler( MainDialogGenerated::OnMenuAbout ), this, m_menuItemAbout->GetId());
```

**Comment by jhasse** [comment]
I see, sorry that it broke for you.

Shouldn't wxWidgets handle this in some way? E.g. simulate that they are coming from the menus.

**Comment by Zenju** [comment]
Maybe, but I couldn't find any official wxWidgets documentation regarding event handling for the "app  menu". In any case, the special handling by wxWidgets seems pretty old (at least 8 years according to this ticket: https://trac.wxwidgets.org/ticket/12092), so it's unlikely wxWidgets is considering the above as a bug on their side.

According to their code it seems all they do is hide (not move) the about, preferences and exit items and create duplicates inside the app menu - without any special consideration to event handling:
https://github.com/wxWidgets/wxWidgets/blob/master/src/osx/menu_osx.cpp#L274

**Comment by jhasse** [comment]
> so it's unlikely wxWidgets is considering the above as a bug on their side.

I wouldn't say that, as dynamic event handling (`Bind`) might not been that common 8 years ago. Can you open a ticket at wxWidgets bug tracker and see what they think?

**Comment by sodevel** [comment]
The advantage of these menu events is that they propagate upwards, it is not necessary to bind at the source, you can bind at the top level element. I can't remember how much this is the official wxWidgets way to handle these events, i just remember that in the past you only had connect for dynamic binding and with that method you could easily create handler methods that have a "broken" this pointer, so connect was only used to bind at the source for events that do not propagate upwards.

I don't know exactly why the handling of menu events was changed in wxFB, in my application code i pretty much bind everything at the top level element and only at the source if i have to, pretty much what wxFB is currently doing for almost all event types.

**Comment by jhasse** [comment]
> I don't know exactly why the handling of menu events was changed in wxFB, in my application code i pretty much bind everything at the top level element and only at the source if i have to, pretty much what wxFB is currently doing for almost all event types.

I need to move the wxMenus into a different window and want that the event handlers still work. Do you know any other way that would be possible?

**Comment by jhasse** [comment]
> With current wxFB you don't have that many options.

Well, it works the way it does now.

Having an extra source field sounds like overkill to me.

**Comment by sodevel** [comment]
Not for the macOS guy :smile:

**Comment by jhasse** [comment]
Yeah, I meant moving wxMenus into different windows (on Linux and Windows).

Regarding macOS: I still think this should be fixed in wxWidgets itself.
